else keyword not correctly parsed

I was writing a script getting the following error:
```bash
(1,1): error CS1525: Unexpected symbol `else' 
```

To verify that the problem is really about the `else` keyword I've reduced it to the following snippet:
```csharp
using System;

var x = true;
if (x) {
	Console.WriteLine("IF {0}", x);
} else {
	Console.WriteLine("ELSE {0}", x);
}
```

I've installed **scriptcs** via `brew` on OS X Yosemite. I can't tell you the version using `sciptcs -Version`, since it fails printing the version.

```bash
➜  temp  brew info scriptcs
scriptcs: stable 0.14.1 (bottled)
Tools to write and execute C#
https://github.com/scriptcs/scriptcs
/usr/local/Cellar/scriptcs/0.14.1 (31 files, 9,9M) *
  Poured from bottle
From: https://github.com/Homebrew/homebrew/blob/master/Library/Formula/scriptcs.rb
==> Dependencies
Recommended: mono ✔
==> Options
--without-mono
	Build without mono support
```

If you need any other info to solve the issue, please tell me.

Is this in the REPL or within a script?

On Thu, Jul 2, 2015 at 10:13 AM Giacomo Stelluti Scala <
notifications@github.com> wrote:

> I was writing a script getting the following error:
>
> (1,1): error CS1525: Unexpected symbol `else'
>
> To verify that the problem is really about the else keyword I've reduced
> it to the following snippet:
>
> using System;
> var x = true;if (x) {
>     Console.WriteLine("IF {0}", x);
> } else {
>     Console.WriteLine("ELSE {0}", x);
> }
>
> I've installed *scriptcs* via brew on OS X Yosemite. I can't tell you the
> version using sciptcs -Version, since it fails printing the version.
>
> ➜  temp  brew info scriptcs
> scriptcs: stable 0.14.1 (bottled)
> Tools to write and execute C#https://github.com/scriptcs/scriptcs
> /usr/local/Cellar/scriptcs/0.14.1 <https://github.com/scriptcs/scriptcs/usr/local/Cellar/scriptcs/0.14.1> (31 files, 9,9M) *
>   Poured from bottle
> From: https://github.com/Homebrew/homebrew/blob/master/Library/Formula/scriptcs.rb
> ==> Dependencies
> Recommended: mono ✔
> ==> Options
> --without-mono
>     Build without mono support
>
> If you need any other info to solve the issue, please tell me.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scriptcs/scriptcs/issues/1060>.
>

I just repro'd this. Looks like it is indeed a bug.

On Thu, Jul 2, 2015 at 10:19 AM Glenn Block <glenn.block@gmail.com> wrote:

> Is this in the REPL or within a script?
>
> On Thu, Jul 2, 2015 at 10:13 AM Giacomo Stelluti Scala <
> notifications@github.com> wrote:
>
>> I was writing a script getting the following error:
>>
>> (1,1): error CS1525: Unexpected symbol `else'
>>
>> To verify that the problem is really about the else keyword I've reduced
>> it to the following snippet:
>>
>> using System;
>> var x = true;if (x) {
>>     Console.WriteLine("IF {0}", x);
>> } else {
>>     Console.WriteLine("ELSE {0}", x);
>> }
>>
>> I've installed *scriptcs* via brew on OS X Yosemite. I can't tell you
>> the version using sciptcs -Version, since it fails printing the version.
>>
>> ➜  temp  brew info scriptcs
>> scriptcs: stable 0.14.1 (bottled)
>> Tools to write and execute C#https://github.com/scriptcs/scriptcs
>> /usr/local/Cellar/scriptcs/0.14.1 <https://github.com/scriptcs/scriptcs/usr/local/Cellar/scriptcs/0.14.1> (31 files, 9,9M) *
>>   Poured from bottle
>> From: https://github.com/Homebrew/homebrew/blob/master/Library/Formula/scriptcs.rb
>> ==> Dependencies
>> Recommended: mono ✔
>> ==> Options
>> --without-mono
>>     Build without mono support
>>
>> If you need any other info to solve the issue, please tell me.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/scriptcs/scriptcs/issues/1060>.
>>
>

"(1,1): error CS1525: Unexpected symbol `else'". I tried also removing
braces, changing to an explicit var. Braces on the same line, braces on
diff lines.

On Thu, Jul 2, 2015 at 10:33 AM Glenn Block <glenn.block@gmail.com> wrote:

> Nope. I tried about 5 diff combos. else is just not recognized
>
> On Thu, Jul 2, 2015 at 10:31 AM Adam Ralph <notifications@github.com>
> wrote:
>
>> Does it work if the script looks like this?
>>
>> using System;
>> var x = true;if (x)
>> {
>>     Console.WriteLine("IF {0}", x);
>> }else
>> {
>>     Console.WriteLine("ELSE {0}", x);
>> }
>>
>> —
>>
>>
>> Reply to this email directly or view it on GitHub
>> <https://github.com/scriptcs/scriptcs/issues/1060#issuecomment-118101657>
>> .
>>
>

--- a/src/ScriptCs.Engine.Mono/Segmenter/Lexer/ScriptLexer.cs
+++ b/src/ScriptCs.Engine.Mono/Segmenter/Lexer/ScriptLexer.cs
@@ -1,4 +1,5 @@
 ﻿using System;
+using System.Collections.Generic;
 using System.IO;
 using System.Text.RegularExpressions;
 
@@ -9,12 +10,22 @@ namespace ScriptCs.Engine.Mono.Segmenter.Lexer
         private int _lastChar = ' ';
         private int _position;
         private readonly StringReader _sr;
+        private readonly Dictionary<string, int> _tokenMap;
 
         public ScriptLexer(string code)
         {
             _position = -1; //inital pos
 
+            _tokenMap = new Dictionary<string, int>
+            {
+                { "if", Token.If },
+                { "else", Token.Else },
+                { "do", Token.Do },
+                { "while", Token.While }
+            };
+
             _sr = new StringReader(code);
+
         }
 
         /// <summary>
@@ -92,13 +103,9 @@ namespace ScriptCs.Engine.Mono.Segmenter.Lexer
                 }
 
                 var token = Token.Identifier;
-                if (identifier.Equals("do"))
+                if (_tokenMap.ContainsKey(identifier))
                 {
-                    token = Token.Do;
-                }
-                else if (identifier.Equals("while"))
-                {
-                    token = Token.While;
+                    token = _tokenMap[identifier];
                 }
 
                 return new LexerResult

--- a/src/ScriptCs.Engine.Mono/Segmenter/Lexer/Token.cs
+++ b/src/ScriptCs.Engine.Mono/Segmenter/Lexer/Token.cs
@@ -8,6 +8,8 @@
         public const int Character = -4;
         public const int Do = -5;
         public const int While = -6;
+        public const int If = -7;
+        public const int Else = -8;
         public const int LeftBracket = '{';
         public const int RightBracket = '}';
         public const int LeftParenthese = '(';

--- a/src/ScriptCs.Engine.Mono/Segmenter/Parser/RegionParser.cs
+++ b/src/ScriptCs.Engine.Mono/Segmenter/Parser/RegionParser.cs
@@ -36,11 +36,20 @@ namespace ScriptCs.Engine.Mono.Segmenter.Parser
                         _current = _lexer.GetToken();
                         if (_current.Token == Token.While)
                         {
-                            var whileRegion = ParseBlock();
-                            var doWhile = region.Combine(whileRegion);
-                            result.Add(doWhile);
+                            var block = ParseBlock();
+                            region = region.Combine(block);
                             _current = _lexer.GetToken();
-                            break;
+                        }
+                        result.Add(region);
+                        break;
+                    case Token.If:
+                        region = ParseBlock();
+                        _current = _lexer.GetToken();
+                        while(_current.Token == Token.Else)
+                        {
+                            var block = ParseBlock();
+                            region = region.Combine(block);
+                            _current = _lexer.GetToken();
                         }
                         result.Add(region);
                         break;

--- a/test/ScriptCs.Engine.Mono.Tests/Segmenter/RegionParserTests.cs
+++ b/test/ScriptCs.Engine.Mono.Tests/Segmenter/RegionParserTests.cs
@@ -175,6 +175,34 @@ namespace ScriptCs.Engine.Mono.Tests.Segmenter
                 region = result[1];
                 Code.Substring(region.Offset, region.Length).ShouldEqual("if (true);");
             }
+
+            [Fact]
+            public void ShouldExtractIfElseAsSingleBlock()
+            {
+                const string Code = "if { } else { }";
+
+                var parser = new RegionParser();
+                var result = parser.Parse(Code);
+
+                result.Count().ShouldEqual(1);
+
+                var region = result[0];
+                Code.Substring(region.Offset, region.Length).ShouldEqual(Code);
+            }
+
+            [Fact]
+            public void ShouldExtractMultipleIfElseAsSingleBlock()
+            {
+                const string Code = "if { } else if { } else if { } else { }";
+
+                var parser = new RegionParser();
+                var result = parser.Parse(Code);
+
+                result.Count().ShouldEqual(1);
+
+                var region = result[0];
+                Code.Substring(region.Offset, region.Length).ShouldEqual(Code);
+            }
         }
 
         public class ParseExpressions

